CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+  // account is verified as an owner of the identifier.
+  hashedPhoneNumber?: string
+  // [Deprecated] Client-specified request timestamp, in milliseconds.
+  // Used for request expiration.
+  timestamp?: number
+  // Client-specified session ID.
+  sessionID?: string
+}
+```
+
+#### Extension for domain-restricted requests
+
+In order to support domain-restricted queries, a new API is proposed:
+
+```typescript
+// POST to /getDomainRestrictedPartialSig
+// Headers:
+//   Content-Type: application/json
+interface GetDomainRestrictedSigRequest {
+  // Domain specification.
+  // Selects the PRF domain and rate limiting rules.
+  domain: Domain
+  // Domain-specific options.
+  // Used for inputs relevant to the domain, but not part of the domain string.
+  // Example: { "authorization": <signature> } for an account-restricted domain.
+  domainOptions?: DomainOptions
+  // Query message. A blinded elliptic curve point encoded in base64.
+  blindedMessage: string
+  // Client-specified session ID.
+  sessionID?: string
+}
+
+interface Domain {
+  // Unique name of the domain. (e.g. "ODIS Password Domain")
+  name: string
+  // Major version number. Allows for backwards incompatible changes.
+  version: number
+  // Arbitrary key-value pairs.
+  // MUST be serializable to EIP-712 encoding.
+  [key: string]: EIP712Value
+}
+
+interface DomainOptions {
+  // Arbitrary key-value pairs.
+  // Valid keys and values depend on the domain specification.
+  // SHOULD be serializable to EIP-712 encoding.
+  [key: string]: EIP712Value
+}
+```
+
+### Domain specifiers
+
+Domain specifiers serve two functions:
+
+- Define the input domain, such that it is uncorrelated to the output in any other domain.
+- Specify the rate limiting function to be used in deciding whether to serve the request.
+
+In order to provide flexibility and support future extensions, the domain specifier is structured
+and allows for the usage of multiple domain-specific rate limiting rules, as support by the ODIS
+service.
+
+In the domain specifiers, two fields MUST be provided: `name` and `version`. Together, these select
+which set of rules the service will apply to the request. Rule sets are included in the ODIS
+implementation, and new rule sets may be added over time. Updating a rule set in a way that would
+change the rules applied to existing domains SHOULD be accompanied by an increase in the `version`
+number. Domain names SHOULD be human-readable. Version numbers SHOULD be whole integers.
+
+With a given `name` and `version`, a number of additional fields MAY be specified to act as
+parameters for the rate limiting rule set (e.g. in a linear backoff rule set, a field may define the
+refresh rate) and to establish the context. In order to ensure deterministic serialization, each
+`name` and `version` MUST correspond to a single type definition.
+
+Changing any field in the domain specifier, results in a new domain with completely unrelated
```

Comment:
Should clients specify the `name` of a domain from a pre-determined set of domain names (like a publicly-available list of rate-limiting schemes that Celo has)? If the `name` is different and doesn't correspond to an existing rate-limiting scheme, would the protocol return some valid output (by just generating a different salt or something along those lines) or should the request just fail altogether?

Reply:
I think this is clarified later down the proposal (in `HTTP Transport`), but maybe it makes sense to clarify earlier on that domain `name`s _must_ come from a set of existing domains when getting feedback from the community?

Reply:
I'll try to clarify that.